nameko/amqp/publish.py

```diff
@@ -0,0 +1,209 @@
+    """ Raised when publisher confirms are enabled and a message could not
+    be routed or persisted """
+    pass
+
+
+@contextmanager
+def get_connection(amqp_uri):
+    conn = Connection(amqp_uri)
+    with connections[conn].acquire(block=True) as connection:
+        yield connection
+
+
+@contextmanager
+def get_producer(amqp_uri, confirms=True):
+    transport_options = {
+        'confirm_publish': confirms
+    }
+    conn = Connection(amqp_uri, transport_options=transport_options)
+
+    with producers[conn].acquire(block=True) as producer:
+        yield producer
+
+
+class Publisher(object):
+    """
+    Utility helper for publishing messages to RabbitMQ.
+    """
+
+    use_confirms = True
+    """
+    Enable `confirms <http://www.rabbitmq.com/confirms.html>`_ for this
+    publisher.
+
+    The publisher will wait for an acknowledgement from the broker that
+    the message was receieved and processed appropriately, and otherwise
+    raise. Confirms have a performance penalty but guarantee that messages
+    aren't lost, for example due to stale connections.
+    """
+
+    delivery_mode = PERSISTENT
+    """
+    Default delivery mode for messages published by this Publisher.
+    """
+
+    mandatory = False
+    """
+    Require `mandatory <https://www.rabbitmq.com/amqp-0-9-1-reference.html
+    #basic.publish.mandatory>`_ delivery for published messages.
+    """
+
+    priority = 0
+    """
+    Priority value for published messages, to be used in conjunction with
+    `consumer priorities <https://www.rabbitmq.com/priority.html>_`.
+    """
+
+    expiration = None
+    """
+    `Per-message TTL <https://www.rabbitmq.com/ttl.html>`_, in milliseconds.
+    """
+
+    serializer = "json"
+    """ Name of the serializer to use when publishing messages.
+
+    Must be registered as a
+    `kombu serializer <http://bit.do/kombu_serialization>`_.
+    """
+
+    compression = None
+    """ Name of the compression to use when publishing messages.
+
+    Must be registered as a
+    `kombu compression utility <http://bit.do/kombu-compression>`_.
+    """
+
+    retry = True
+    """
+    Enable automatic retries when publishing a message that fails due
+    to a connection error.
+
+    Retries according to :attr:`self.retry_policy`.
+    """
+
+    retry_policy = DEFAULT_RETRY_POLICY
+    """
+    Policy to apply when retrying message publishes, if requested.
+
+    See :attr:`self.retry`.
+    """
+
+    declare = []
+    """
+    Kombu :class:`~kombu.messaging.Queue` or :class:`~kombu.messaging.Exchange`
+    objects to (re)declare before publishing a message.
+    """
+
+    def __init__(
+        self, amqp_uri, use_confirms=None, serializer=None, compression=None,
+        delivery_mode=None, mandatory=None, priority=None, expiration=None,
+        declare=None, retry=None, retry_policy=None, **publish_kwargs
+    ):
+        self.amqp_uri = amqp_uri
+
+        # publish confirms
+        if use_confirms is not None:
```

Comment:
Several of them have truthy defaults, so it's hard to compare whether a new value has been provided. This is a little more verbose but consistent for every param. I also like that the default values sit next to the docs for each option.

A slightly less verbose option could be the following:

``` python
def __init__(self, amqp_uri, **kwargs):
    self.amqp_uri = amqp_uri

    self.use_confirms = kwargs.pop('use_confirms', self.use_confirms)
    ...
```

Slight disadvantage here is that the sphinx docs would only show `**kwargs`.

Re: adding too many ways to change settings. Maybe. This does allow subclassing to override defaults as well as at instantiation (and use-time). Subclassing used to be the only way to change these things previously though, and we encourage subclassing for other reasons.